backend/copr_backend/background_worker_build.py

```diff
@@ -404,20 +404,34 @@ def _proctitle(self, text):
 
     def _check_build_interrupted(self):
         """
-        Should we interrupet a running worker?
+        Should we interrupt a running worker?
         """
         return (self._check_failed_resalloc_ticket()
-                or self._cancel_task_check_request())
+                or self._cancel_task_check_request()
+                or self._build_timeouted())
 
     @ttl_cache(ttl=10*60)
     def _check_failed_resalloc_ticket(self):
         """
         Did the resalloc ticket fail?
         """
         self.host.ticket.collect()
-        self.log.info("Status: %s", self.host.ticket.failed)
+        self.log.info("Failed resalloc ticket: %s", self.host.ticket.failed)
         return self.host.ticket.failed
 
+    def _build_timeouted(self):
+        """
+        When build timeouts, it should be handled by `copr-rpmbuild` and the
+        builder machine should mark itself as finished. When it fails to do so,
+        we have this fail-safe to know if a build timeouted and should be
+        terminated.
+        """
+        # Wait some time (1 hour) after the configured timeout for the builder
+        # to terminate itself.
+        timestamp = self.job.started_on + self.job.timeout + 60 * 60
+        limit = datetime.fromtimestamp(timestamp)
+        return datetime.now() > limit
```

Comment:
I think this is correct approach - killing the builder from backend when it timeouts. Handling the timeout should be done by layer above (read as - should be handled by the caller) in case of any bug inside the inner layer (builder here) which will result in stuck builder thus it never timeouts. 

Take CPU scheduler as example - the process has some time given and after that, the resource is taken from it by outer layer - the process does not decide when it want's to stop. This way the process could maliciously stole the CPU resource for itself. In our case it won't probably be maliciously but by mistake.

Considering this above, I am more inclined to handle timeout only by backend (at the timeout time, without waiting another hour) and drop the timeout handlind at builder level. I'll give it more thought after sleep tho.

Reply:
The rpmbuild logic implements a "gently" timeout; I don't think it is desirable to remove that logic (or at even spend the time on its removal).  The PR here seems like a valid implementation. +1